### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by GregGovit**
J'ai implémenté SMORMS3 et une version légèrement modifiée de celui-ci SMORMS3Mod qui me semble meilleure ;-). Les 2 ont l'air de converger mieux et plus vite que RMSPROP (en tout cas sur les tests que je viens de faire). Je laisse tourner cette nuit et je t'enverrai les résultats demain.

Pour le tester, il faut mettre dans ton script "train.py" les lignes suivantes :

from pyannote.audio.embedding.optimizers import SMORMS3Mod
opti = SMORMS3Mod()

from pyannote.audio.embedding.base import SequenceEmbedding
embedding = SequenceEmbedding(loss=loss, optimizer=opti, log_dir=LOG_DIR)

### CODE REVIEW

**Comment by hbredin** [comment]
Je m'apprête à tester tout ça 
Tu conseilles d'utiliser les paramètres par défaut ?

**Comment by GregGovit** [comment]
Oui, il n’y a rien à faire ça marche tout seul. Il faut utiliser SMORMS3Mod() (qu’on pourrait peut-être renommer de façon plus sexy, genre stabilizedSMORMS3 ;-) ).
Voici les résultats sur le test avec une architecture RNN légèrement modifiée (Tristounette) :

EER = 22.14%
EER = 20.25%
EER = 16.37%
EER = 16.39%
EER = 15.56%
EER = 14.42%
EER = 14.73%
EER = 14.31%
EER = 14.37%
EER = 13.85%
EER = 14.04%
EER = 14.30%
EER = 14.07%
EER = 13.98%
EER = 14.15%
EER = 14.12%
EER = 14.05%
EER = 13.86%
EER = 13.86%
EER = 13.81%
EER = 13.81%
EER = 13.61%
EER = 13.73%
EER = 13.91%
EER = 13.89%
EER = 13.59%
EER = 13.50%
EER = 13.66%
EER = 13.45%
EER = 13.65%
EER = 13.53%
EER = 13.52%
EER = 13.49%
EER = 13.66%
EER = 13.54%
EER = 13.53%
EER = 13.51%
EER = 13.55%
EER = 13.71%
EER = 13.56%
EER = 13.48%
EER = 13.46%
EER = 13.45%

Le répertoire avec les figures et les sauvegardes se trouvent ici : /vol/lid/gelly/keras/test_mod_tristounette/

La convergence me semble plus lente mais plus stable qu’avec Tristounet.

Grégory

**Comment by hbredin** [comment]
Je confirme que les SMORMS3 (originale et modifiée) fonctionnent mieux que RMSProp sur TristouNet.

En revanche, sur TristouNet en 3d (output_dim=3 au lieu de output_dim=16), le comportement n'est pas top. Voir l'évolution du EER époque après époque ci-dessous. Ca descend jusqu'à EER=23% pour l'époque 6, puis ça explose. C'est probablement que 3 dimension, ce n'est pas suffisant mais je voulais te faire part de ce comportement.

1 30.25%
2 30.18%
3 30.96%
4 29.81%
5 28.89%
6 23.54%
7 31.96%
8 28.80%
9 29.65%
10 31.18%
11 27.17%
12 29.65%

**Comment by GregGovit** [comment]
Cool.
Le cas 3D, c’est avec la version modifiée ou originale ?

File: pyannote/audio/embedding/models.py [hunk 1]
```diff
@@ -172,3 +174,134 @@ def __call__(self, input_shape):
...
+                return_sequences = i+1 < n_lstm
+            elif self.pooling == 'average':
+                return_sequences = True
+            else:
+                raise NotImplementedError(
+                    'unknown "{pooling}" pooling'.format(pooling=self.pooling))
+
+            if i:
+                # all but first LSTM
+                if self.bidirectional:
```
Comment (hbredin) [inline pyannote/audio/embedding/models.py]: While we are at it, I'd rather go with something like:

```python
lstm = LSTM(name='lstm_{i:d}'.format(i=i), ...)
if self.bidirectional:
    lstm = Bidirectional(lstm)
```

File: pyannote/audio/embedding/models.py [hunk 2]
```diff
@@ -172,3 +174,134 @@ def __call__(self, input_shape):
                             name="embedding_output")(x)
 
         return Model(input=inputs, output=embeddings)
+
+class TristouNette(object):
```
Comment (hbredin) [inline pyannote/audio/embedding/models.py]: Sorry for being a pain in the $*^%$ but I'd rather have two "unitary" pull requests: one dealing with models, one dealing with optimizers.

**Comment by hbredin** [comment]
Closing, as this is now superseded by latest commits in `develop` branch.
